sources/engine/Stride.Input/Windows/RawInput/RawInput.cs

```diff
@@ -0,0 +1,144 @@
+            public int RawButtons;
+            public int LastX;
+            public int LastY;
+            public int ExtraInformation;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct RawKeyboard
+        {
+            public ushort MakeCode;
+            public ushort Flags;
+            public ushort Reserved;
+            public ushort VKey;
+            public uint Message;
+            public uint ExtraInformation;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct RawHid
+        {
+            public uint dwSizeHid;
+            public uint dwCount;
+            public int bRawData;
+        }
+
+        [StructLayout(LayoutKind.Explicit)]
+        public struct RawInputDataInner
+        {
+            [FieldOffset(0)]
+            public RawMouse Mouse;
+
+            [FieldOffset(0)]
+            public RawKeyboard Keyboard;
+
+            [FieldOffset(0)]
+            public RawHid Hid;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        public struct RawInputData
+        {
+            public RawInputHeader header;
+            public RawInputDataInner data;
+        }
+
+        public static unsafe byte[] GetHIDRawData(ref RawHid rawHid)
+        {
+            var byteCount = rawHid.dwCount * rawHid.dwSizeHid;
+            var result = new byte[byteCount];
+            fixed (byte* dest = result)
+            {
+                fixed (int* src = &rawHid.bRawData)
+                {
+                    Unsafe.CopyBlockUnaligned(dest, src, byteCount);
+                }
+            }
+            return result;
+        }
+
+        private static unsafe bool RegisterDevice(RawInputDevice device)
+        {
+            var devices = new RawInputDevice[1] { device };
+            fixed (void* ptr = devices)
+            {
+                return Win32.RegisterRawInputDevices(ptr, 1, (uint)sizeof(RawInputDevice));
+            }
+        }
+
+        public static bool RegisterDevice(UsagePage usagePage, UsageId usageId, ModeFlags flags, IntPtr target)
+        {
+            var device = new RawInputDevice
+            {
+                UsagePage = (ushort)usagePage,
+                Usage = (ushort)usageId,
+                Flags = (uint)flags,
+                Target = target
+            };
+            return RegisterDevice(device);
+        }
+
+        public static unsafe RawInputData* GetRawInputData(IntPtr lParam)
+        {
+            uint cbSize = 0;
+            Win32.GetRawInputData(lParam, (uint)RawInputDataType.RID_INPUT, IntPtr.Zero, ref cbSize, (uint)sizeof(RawInputHeader));
+            if (cbSize == 0)
+            {
+                return default;
+            }
+            var buffer = stackalloc byte[(int)cbSize];
+            var count = Win32.GetRawInputData(lParam, (uint)RawInputDataType.RID_INPUT, (IntPtr)buffer, ref cbSize, (uint)sizeof(RawInputHeader));
+            return (RawInputData*)buffer;
```

Comment:
Wouldn't this cause problems? Basically you are allocating a buffer on the stack and returning a pointer to that buffer. After exiting the function that buffer is out of scope and can be potentially overwritten by other data.

Reply:
ooh, yeah. That is a very good point. I had it different before but apparently screwed it up later. Good catch.

Reply:
This is now fixed :)